[quant][dbr] Add support for functional conv variants

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71795

Summary: This commit expands the API coverage of functional conv
ops in DBR quantization to F.conv1d, F.conv2d, and F.conv3d.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

andrewor14 committed:
DBR Quantization: Add support for functional conv variants

Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/_dbr/mappings.py:
```diff
@@ -23,7 +23,12 @@
+    F.conv1d: torch.ops.quantized.conv1d,
     F.conv2d: torch.ops.quantized.conv2d,
+    F.conv3d: torch.ops.quantized.conv3d,
+    F.conv_transpose1d: torch.ops.quantized.conv_transpose1d,
```
I see that the tests cover `F.conv1d` and `F.conv3d`, are the transpose variants tested as well?

andrewor14 replied:
Yes, I'm adding the tests now. They didn't work out of the box and I need to look into why.

vkuzo commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -118,40 +118,49 @@ class TestQuantizeDBRIndividualOps(QuantizeDBRTestCase):
+        for dim in range(1, 4):
+            m = M(dim).eval()
+            qconfig = torch.quantization.default_qconfig
+            data = self.img_data_dict[dim][0][0]
```
nit: could we use something like `torch.randn` instead of `self.img_data_dict`?  this is a personal preference, it's easier to use these tests for debugging if the all the information is specified inline and we don't have to navigate to utils files and base classes to understand the structure of the input data

vkuzo replied:
just curious, thoughts about my comment above?

andrewor14 replied:
Oh sorry I missed this one

vkuzo reviewed (commented):
thanks for working on this! Looks good, I had some minor comments inline

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

[ghstack-poisoned]

vkuzo commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -118,40 +118,81 @@ class TestQuantizeDBRIndividualOps(QuantizeDBRTestCase):
-                x = F.conv2d(
-                    x, self.weight2d, self.bias2d, self.stride2d, self.padding2d,
-                    self.dilation2d, self.groups)
+                x = convs[dim](
```
can we assign it in the constructor and call `self.conv_fun` in the forward? it would just be easier to read the code this way

vkuzo commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -118,40 +118,81 @@ class TestQuantizeDBRIndividualOps(QuantizeDBRTestCase):
+                    self.output_padding, self.groups, self.dilation)
+                return x
+
+        for dim in range(1, 4):
+            weights = {
+                1: torch.randn(1, 3, 10),
+                2: torch.randn(1, 3, 10, 10),
+                3: torch.randn(1, 3, 5, 5, 5)
+            }
+            data = dict([(k, torch.transpose(v, 0, 1)) for k, v in weights.items()])
```
can we just spell out weights and data, without using a formula?  A good thing to optimize for here is readability, lines of code and code reuse in this case don't seem as important.  Ideally the code is dead simple to understand for someone who is trying to fix something, copy-pasting a few lines is okay.

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 commented:
Oops sorry something is wrong with the tests. Let me look into it first.

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/_dbr/mappings.py:
```diff
@@ -110,6 +125,27 @@
+    F.conv1d,
+    F.conv2d,
+    F.conv3d,
+]).union(conv_transpose_ops)
```
nit: might be ambiguous to say that "conv transpose ops are conv ops", there are some arguments for and some against.  What do you think about keeping them separate, to make it more clear?

andrewor14 replied:
Yes, I went back and forth on this too. For the most part both conv and conv transpose go through the same code paths, so there will be many places where we have to say `if op in conv_ops or op in conv_transpose_ops`, which is a bit verbose IMO. On the other hand, keeping them separate might be clearer as you suggested. I don't have a strong preference one way or the other.

vkuzo replied:
IMO usually the way which is clear and easy to read is preferrable in cases like this.

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 committed:
Update on "DBR Quantization: Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

vkuzo commented:
code looks good, can we fix the tests before landing?

andrewor14 committed:
Update on "[quant][dbr] Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization from F.conv2d to all conv variants.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 commented:
As discussed offline, I'm removing support for functional conv transpose for now because that is not supported in FX yet either. The latest changes add support for only F.conv1d and F.conv3d.

andrewor14 commented:
As discussed offline, `test_conv_transpose_functional` was failing because functional conv transpose is not supported in FX yet either. I will remove support for functional conv transpose in this PR for now (along with the failing test) and leave it as future work until that functionality is added in FX. The latest changes will add support for only F.conv1d and F.conv3d.

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33907099).

andrewor14 committed:
Update on "[quant][dbr] Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization to F.conv1d, F.conv2d, and F.conv3d.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][dbr] Add support for functional conv variants"


Summary: This commit expands the API coverage of functional conv
ops in DBR quantization to F.conv1d, F.conv2d, and F.conv3d.

Test Plan: python test/test_quantization.py TestQuantizeDBRIndividualOps.test_conv_functional

Reviewers: vkuzo

Subscribers: vkuzo, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/70620

Differential Revision: [D33907099](https://our.internmc.facebook.com/intern/diff/D33907099)

[ghstack-poisoned]
